sled-hardware/src/disk.rs

```diff
@@ -503,26 +520,50 @@ impl Disk {
                 DiskError::IoError { path: keyfile.path().0.clone(), error }
             })?;
 
-            result?
-        } else {
-            false
+            result?;
         };
 
         for dataset in datasets.into_iter() {
             let mountpoint = zpool_name.dataset_mountpoint(dataset.name);
             let name = &format!("{}/{}", zpool_name, dataset.name);
 
-            if dataset.wipe && newly_mounted_crypt {
-                info!(log, "Automatically destroying dataset {}", name);
-                Zfs::destroy_dataset(name).or_else(|err| {
-                    // If we can't find the dataset, that's fine -- it might
-                    // not have been formatted yet.
-                    if let DestroyDatasetErrorVariant::NotFound = err.err {
-                        Ok(())
-                    } else {
-                        Err(err)
+            // Use a value that's alive for the duration of this sled agent
+            // to answer the question: should we wipe this disk, or have
+            // we seen it before?
+            //
+            // If this value comes from a prior iteration of the sled agent,
+            // we opt to remove the corresponding dataset.
+            static AGENT_LOCAL_VALUE: OnceLock<String> = OnceLock::new();
+            let agent_local_value = AGENT_LOCAL_VALUE.get_or_init(|| {
+                Alphanumeric.sample_string(&mut rand::thread_rng(), 12)
```

Comment:
I'm only avoiding a UUID to identify "this is random junk" -- it doesn't matter what it is, as long as it's different each time the sled agent boots.

Reply:
I am definitely not certain whether this matters, but if we want the same amount of entropy as a UUIDv4, we want a 21-character long string here (log_2(62^21) >= 122).

Reply:
I don't *think* we care about entropy quite as much -- it's not a global space, it's pretty much just a single value that's sled local (so risk of collision is significantly lower) but I can bump the size regardless.